graylog2-server/src/main/java/org/graylog2/shared/rest/resources/ProxiedResource.java

```diff
@@ -135,37 +204,80 @@ protected ProxiedResource(@Context HttpHeaders httpHeaders,
                 .findFirst()
                 .orElseThrow(() -> new IllegalStateException("No active master node found"));
 
-        final RemoteInterfaceType remoteInterfaceType = remoteInterfaceProvider.apply(masterNode.getNodeId())
-                .orElseThrow(() -> new IllegalStateException("Master node " + masterNode.getNodeId() + " not found"));
+        return doNodeApiCall(masterNode.getNodeId(), remoteInterfaceProvider, remoteInterfaceFunction, Function.identity());
+    }
+
+    private  <RemoteInterfaceType, RemoteCallResponseType, FinalResponseType> MasterResponse<FinalResponseType> doNodeApiCall(
+            String nodeId,
+            Function<String, Optional<RemoteInterfaceType>> remoteInterfaceProvider,
+            Function<RemoteInterfaceType, Call<RemoteCallResponseType>> remoteInterfaceFunction,
+            Function<RemoteCallResponseType, FinalResponseType> transformer
+    ) throws IOException {
+        final RemoteInterfaceType remoteInterfaceType = remoteInterfaceProvider.apply(nodeId)
+                .orElseThrow(() -> new IllegalStateException("Node " + nodeId + " not found"));
 
         final Call<RemoteCallResponseType> call = remoteInterfaceFunction.apply(remoteInterfaceType);
         final Response<RemoteCallResponseType> response = call.execute();
 
         final byte[] errorBody = response.errorBody() == null ? null : response.errorBody().bytes();
 
-        return MasterResponse.create(response.isSuccessful(), response.code(), response.body(), errorBody);
+        return MasterResponse.create(response.isSuccessful(), response.code(), transformer.apply(response.body()), errorBody);
+    }
+
+    /**
+     * This wrapper is intended to provide additional server error information
+     * if something went wrong beyond the actual API HTTP call.
+     */
+    @AutoValue
+    public static abstract class CallResult<ResponseType> {
+
+        @JsonProperty("success")
+        public abstract boolean isSuccess();
+
+        @JsonProperty("server_error_message")
+        @Nullable
+        public abstract String serverErrorMessage();
+
+        @JsonProperty("response")
+        @Nullable
+        public abstract MasterResponse<ResponseType> response();
```

Comment:
I would rename `MasterResponse` to something like `NodeResponse`

It is not just from master nodes.

Reply:
Yes. Initially it was called `NodeResponse`, but later I had to revert it not to break plugins using this class.